clash-ghc/src-ghc/Clash/GHC/GHC2Core.hs

```diff
@@ -896,6 +916,44 @@ modNameM n = do
   let moduleNm = moduleName module_
   return (pack (moduleNameString moduleNm))
 
+-- | Given the type:
+--
+-- @
+--     forall dom a0 a1 .. aN
+--   . Signal dom (a0, a1, .., aN)
+--  -> (Signal dom a0, Signal dom a1, .., Signal dom aN)
+-- @
+--
+-- or the type
+--
+-- @
+--     forall dom a0 a1 .. aN
+--   . (Signal dom a0, Signal dom a1, .., Signal dom aN)
+--  -> Signal dom (a0, a1, .., aN)
+-- @
+--
+-- Generate the term:
+--
+-- @/\dom. /\a0. /\a1. .. /\aN. \x -> x@
+--
+-- In other words: treat "bundle" and "unbundle" primitives as id.
+--
+bundleUnbundleTerm :: Int -> C.Type -> C.Term
+bundleUnbundleTerm nTyVarsExpected = go []
```

Comment:
If you'd write it

```haskell
bundleUnbundleTerm nTyVarsExpected ty = go (-1) ty
 where
  go !n (C.ForAllTy tv typ) = C.TyLam tv (go (n+1) typ)
  go n (C.tyView -> C.FunTy argTy _resTy) =
     if n == nTyVarsExpected then
       let sigName = C.mkLocalId argTy (C.mkUnsafeSystemName "c$s" 0) in
       C.Lam sigName (C.Var sigName)
     else
       error ($(curLoc) ++ show n ++ " vs " ++ show nTyVarsExpected)
  go _ _ = error ($(curLoc) ++ C.showPpr ty)
```

you don't need the `reverse`, and you don't need to pass in the `(n+1)` in the call to `bundleUnbundleTerm`

Reply:
Ah that's clever. Done!